prov/util/src/util_mr_cache.c

```diff
@@ -245,6 +258,12 @@ util_mr_cache_create(struct ofi_mr_cache *cache, const struct ofi_mr_info *info,
 	if (ret)
 		goto free;
 
+#ifdef HAVE_LIBCUDA
+	ret = ofi_cuda_get_buf_id(*entry, &info->iov);
+	if (ret < 0)
+		goto free;
+#endif
```

Comment:
I prefer to avoid #ifdef in C code, unless it surrounds the entire function.  I'd rather the function always exist, and just do nothing if cuda isn't available.

On this line, how do we know that the buffer is from cuda?  Are we forced to blindly call this for all buffers?

I would consider creating a new function: ofi_get_buf_hmem_id() (? names are hard), which can call cuda and other gpu libraries, assuming that they all support some sort of buffer identification mechanism.

Reply:
>I would consider creating a new function: ofi_get_buf_hmem_id() (? names are hard), which can call cuda and other gpu libraries, assuming that they all support some sort of buffer identification mechanism.
AMD GPUs would not use this. AMD ROCr allows users to register a GPU buffer deallocation callback.

Reply:
> I'd rather the function always exist, and just do nothing if cuda isn't available.

The functions in ofi_cuda.h are all conditionally defined, but makes sense, we can change those to stubs if the CUDA runtime is not available.

> On this line, how do we know that the buffer is from cuda? Are we forced to blindly call this for all buffers?

We'll need a field in the cache entry to tell what iface type was used for the allocation. I rushed this change to do some functional tests with the tag check and did not "productize" it :)

> ofi_get_buf_hmem_id() (? names are hard), which can call cuda and other gpu libraries, assuming that they all support some sort of buffer identification mechanism.

We could. Did not want to assume all hmem devices would provide this. Looks like @iziemba already has an example. If we did add a buffer ID field to fi_mr_attr, based on the hmem type, ofi_get_buf_hmem_id could validate on devices where that mechanism is available.

Reply:
Sigh, of course the APIs are different.  I think the ofi_mr_cache_entry_usable() is still the right place for abstracting away the validation checks.

Maybe here we need something even more generic, like ofi_mr_entry_hmem_init(), which can determine what type of memory we're dealing with (if unknown), which device it belongs to, what it's buffer id is, and/or any other random thing that we'll need later in order to validate the buffer and release it when ready.

Maybe setting up a cache callback, similar to add_region(), would help?  We could also restrict each cache to only supporting host memory plus one hmem type as a first step.  Because other than a test app, is any app really going to use GPUs from different vendors at the same time?